### ISSUE TITLE
Very poor performance using the unique function 

### ISSUE THREAD

**Issue by DrDryg** [pytorch/pytorch#107098]
### 🐛 Describe the bug

When using the unique function it is very slow and uses a lot of memory. I run the following script to do the benchmarking for time, and memory was monitored using the Task manager in windows so those numbers can be a bit approximate.

```
import numpy as np
import torch
from time import time

shape = (1000, 1000, 32)
xt = torch.randn(shape, dtype=torch.float32)
xn = np.random.randn(*shape).astype(dtype=np.float32)


def np1(x):
    t1 = time()
    u = np.unique(x)
    t2 = time()
    print(f"Numpy just sort: {t2-t1} s")


def np2(x):
    t1 = time()
    u, idx, inv, c = np.unique(x, return_index=True, return_inverse=True, return_counts=True)
    t2 = time()
    print(f"Numpy sort + indexes: {t2-t1} s")


def torch1(x: torch.Tensor):
    t1 = time()
    u = x.unique()
    t2 = time()
    print(f"Torch just sort: {t2-t1} s")


def torch2(x: torch.Tensor):
    t1 = time()
    u, idx, c = x.unique(return_inverse=True, return_counts=True)
    t2 = time()
    print(f"Torch sort + indexes: {t2-t1} s")


np1(xn)
np2(xn)
torch1(xt)
torch2(xt)
```

This resulted in the output:

Numpy just sort: 2.6931326389312744 s
Numpy sort + indexes: 6.083210229873657 s
Torch just sort: 16.39222812652588 s
Torch sort + indexes: 43.61249351501465 s

The memory investigation resulted in the following. (Note that the baseline of memory usage was 400MB): 
Numpy just sort: 600 - 400 MB = 200 MB above baseline
Numpy sort + indexes: 1500MB-400MB = 1100MB over baseline
Torch just sort: 1800MB-400MB = 1400MB over baseline.
Torch sort + indexes: 4400MB-400MB = 4000MB over baseline.

The "just sort" methods returned data of size 128MB and the "sort + indexes" returned data of size 640MB. So torch uses several times the amount of memory and takes about three times as long to run. This is for CPU since that is what I mainly is interested in. I have not looked at the issue for GPU.

Actually Im most interested in libtorch rather than pytorch. There I saw similar issues but with even higher memory consumption. I also noted a weird thing. Calling `u, idx, c = x.unique()` used about 20 GB of memory, before it gave an exception during the unpacking.

### Versions

PyTorch version: 2.0.1+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.26.0-rc5
Libc version: N/A

Python version: 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: False
CUDA runtime version: 11.4.100
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Laptop GPU
Nvidia driver version: 527.99
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU: Ryzen 7 5800H 8 core
Memory: 32GB

Revision=20480

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[conda] Could not collect

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by mingfeima** [comment]
Currently `unique` is implemented in a different way from Numpy. PyTorch uses `hash table` and NumPy uses `sort + masked_select`: https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323 We can optimize pytorch implementation in a similar algo (and parallel the masked select as well).



Just one thing: [NumPy unique has an argument `equal_nan`](https://numpy.org/doc/stable/reference/generated/numpy.unique.html) while pytorch treats NaNs as different values. Do we need to align the NumPy behavior or just stick to the current PyTorch's behavior ? @colesbury, do you have any comments ？

If you guys have concern to change the behavior of  `unique` with NaN now, i can leave a `optimized` phony path with argument of `equal_nan=false` in the kernel. And you may easily enable it whenever it feels right to do so. Kind of like:
```C++
std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
    const Tensor& self,
    const bool sorted,
    const bool return_inverse,
    const bool return_counts,
    const bool equal_nan=false) {

if (equal_nan) {
  // this is NumPy's behavior
  kernel_treat_nan_as_equal();
} else {
  // this is current PyTorch's behavior
  kernel_treat_nan_as_different();
}
}
```

**Comment by mingfeima** [comment]
@DrDryg Hi, I have made an initial version here https://github.com/pytorch/pytorch/pull/107846. The algo is identical to numpy's `unique` but with more optimized parallel loops, as you can see `inverse index` and `counts` calculation is almost free launch, only takes very small amount of additional time.

Below is the performance before/after the patch (since int and float have different implementation on sort, i tested both int64 and float32). Test machine is Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz , each socket has 20 cores and the data below are gathered with single socket (20 cores).

```shell
### before (dtype int64)
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s

### after (dtype int64)
Numpy just sort: 0.4126708507537842 s
Numpy sort + indexes: 6.3781516551971436 s
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s

### before (float32)
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s

### after (float32)
Numpy just sort: 0.4073069095611572 s
Numpy sort + indexes: 6.362815856933594 s
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

Note that we don't have parallel sort on float when input is 1-dimensional tensor ( int tensor sort is parallelled with radix sort), so float unqiue on torch is still slower than numpy, if this is fixed, the performance becomes:
```shell
### after (float32 with parallel sort)
Numpy just sort: 0.4129054546356201 s
Numpy sort + indexes: 6.40191912651062 s
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

The parallel sort enabling on CPU with 1-dimensional input will be done in separate PRs.
